picoquic/picosocks.c

```diff
@@ -240,6 +240,21 @@ int picoquic_socket_set_ecn_options(SOCKET_TYPE sd, int af, int * recv_set, int
     return ret;
 }
 
+int picoquic_socket_set_pmtud_options(SOCKET_TYPE sd, int af)
+{
+    int ret = 0;
+#if (defined(IP_MTU_DISCOVER) || defined(IPV6_MTU_DISCOVER)) && defined(IP_PMTUDISC_PROBE)
+    int val = IP_PMTUDISC_PROBE;
+    if (af == AF_INET6) {
+        ret = setsockopt(sd, IPPROTO_IPV6, IPV6_MTU_DISCOVER, &val, sizeof(int));
+    }
+    else {
+        ret = setsockopt(sd, IPPROTO_IP, IP_MTU_DISCOVER, &val, sizeof(int));
+    }
+#endif
+    return ret;
```

Comment:
This does not work. If the `#if` test is not successful, the function becomes empty, the variables `sd` and `af` are not used, and the compiler creates a warning treated as error:
```
2>c:\projects\picoquic\picoquic\picosocks.c(243): error C2220: warning treated as error - no 'object' file generated [C:\projects\picoquic\picoquic\picoquic.vcxproj]
     2>c:\projects\picoquic\picoquic\picosocks.c(243): warning C4100: 'af': unreferenced formal parameter [C:\projects\picoquic\picoquic\picoquic.vcxproj]
     2>c:\projects\picoquic\picoquic\picosocks.c(243): warning C4100: 'sd': unreferenced formal parameter [C:\projects\picoquic\picoquic\picoquic.vcxproj]
```
Also, the test says `(defined(IP_MTU_DISCOVER) || defined(IPV6_MTU_DISCOVER))`, but then you call both the IPV6 and IPv4 variants. What if only one of those is defined?

Reply:
Since the problem only seems to appear on Windows, one way to solve this would be:
```
    int ret = 0;
#if (defined(IP_MTU_DISCOVER) || defined(IPV6_MTU_DISCOVER)) && defined(IP_PMTUDISC_PROBE)
    int val = IP_PMTUDISC_PROBE;
    if (af == AF_INET6) {
        ret = setsockopt(sd, IPPROTO_IPV6, IPV6_MTU_DISCOVER, &val, sizeof(int));
    }
    else {
        ret = setsockopt(sd, IPPROTO_IP, IP_MTU_DISCOVER, &val, sizeof(int));
    }
#else
#ifdef _WINDOWS
    UNREFERENCED_PARAMETER(af);
    UNREFERENCED_PARAMETER(sd);
#endif
#endif
```

Reply:
> This does not work. If the #if test is not successful, the function becomes empty, the variables sd and af are not used, and the compiler creates a warning treated as error:

Oh, right. That's ugly.

> Also, the test says (defined(IP_MTU_DISCOVER) || defined(IPV6_MTU_DISCOVER)), but then you call both the IPV6 and IPv4 variants. What if only one of those is defined?

That would be a problem. I tried to save lines. If you don't mind to have more if statements, I would write:

```
#ifndef __linux__
    UNREFERENCED_PARAMETER(sd);
    UNREFERENCED_PARAMETER(af);
    return 0;
#else
    if (af == AF_INET6) {
#if defined(IPV6_MTU_DISCOVER) && defined(IP_PMTUDISC_PROBE)
        int val = IP_PMTUDISC_PROBE;
        return setsockopt(sd, IPPROTO_IPV6, IPV6_MTU_DISCOVER, &val, sizeof(int));
#else
        UNREFERENCED_PARAMETER(sd);
        return -1;
#endif
    }
    else {
#if defined(IP_MTU_DISCOVER) && defined(IP_PMTUDISC_PROBE)
        int val = IP_PMTUDISC_PROBE;
        return setsockopt(sd, IPPROTO_IP, IP_MTU_DISCOVER, &val, sizeof(int));
#else
        UNREFERENCED_PARAMETER(sd);
        return -1;
#endif
    }
#endif /* #ifndef __linux__ */
```

> Since the problem only seems to appear on Windows

I believe IP_MTU_DISCOVER is a special Linux thing.

Reply:
The code compiles on MacOS, according to the tests reports, but I don't know whether it executes correctly. If we don't care about MAC OS, then we know that on Linux both IP_MTU_DISCOVER and IPV6_MTU_DISCOVER will be defined -- or not, depending on version. So without losing generality, we can simplify the code further. Something like:
```
    int ret = 0;
#if  defined __linux && defined(IP_MTU_DISCOVER) && defined(IPV6_MTU_DISCOVER) && defined(IP_PMTUDISC_PROBE)
    int val = IP_PMTUDISC_PROBE;
    if (af == AF_INET6) {
        ret = setsockopt(sd, IPPROTO_IPV6, IPV6_MTU_DISCOVER, &val, sizeof(int));
    }
    else {
        ret = setsockopt(sd, IPPROTO_IP, IP_MTU_DISCOVER, &val, sizeof(int));
    }
#else
    UNREFERENCED_PARAMETER(af);
    UNREFERENCED_PARAMETER(sd);
#endif  /* #ifndef __linux__ && ... */
```
That version would be much shorter.

Reply:
Done